proxy.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+)
+
+type proxy struct {
+	bootstrapData *bootstrapData
+	reverseProxy  *httputil.ReverseProxy
+	readClusters  bool
+	readListeners bool
+	readEndpoints map[string]struct{}
+}
+
+func newProxy(upstreamURLRaw string, bootstrapData *bootstrapData) (*proxy, error) {
+	upstreamURL, err := url.Parse(upstreamURLRaw)
+	if err != nil {
+		return nil, err
+	}
+
+	return &proxy{
+		bootstrapData: bootstrapData,
+		reverseProxy:  httputil.NewSingleHostReverseProxy(upstreamURL),
+		readEndpoints: make(map[string]struct{}),
+	}, nil
+}
+
+func (p *proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	switch req.URL.Path {
+	case "/v2/discovery:endpoints":
+		p.handleEDS(w, req)
+	case "/v2/discovery:listeners":
+		p.handleLDS(w, req)
+	case "/v2/discovery:clusters":
+		p.handleCDS(w, req)
+	case "/healthz":
+		if len(p.readEndpoints) == len(p.bootstrapData.Endpoints) {
+			http.Error(w, "ok", 200)
+		} else {
+			http.Error(w, "bootstrapping", 500)
+		}
+	default:
+		http.Error(w, "not found", 404)
+	}
+}
+
+func (p *proxy) handleEDS(w http.ResponseWriter, req *http.Request) {
+	if len(p.readEndpoints) == len(p.bootstrapData.Endpoints) {
+		p.reverseProxy.ServeHTTP(w, req)
+		return
+	}
+
+	if req.Method != "POST" {
+		http.Error(w, "method not allowed", 405)
+		return
+	}
+
+	dr, err := readDiscoveryRequest(req)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	if len(dr.ResourceNames) != 1 {
+		http.Error(w, "must have 1 resource_names", 400)
+		return
+	}
+
+	log.Printf("serving request for '%s' endpoints from bootstrap data", dr.ResourceNames[0])
+	p.readEndpoints[dr.ResourceNames[0]] = struct{}{}
+	w.Write(p.bootstrapData.Endpoints[dr.ResourceNames[0]])
```

Comment:
We can probably just null this out after writing it out, right?

Then use this instead of the `readEndpoints` map. It'd also be lighter on memory since we can just evict these. If we're under the assumption they're each only read once, we can just leverage if they are null or not. Or popping the keys out of bootstrapData, etc.

I think we can make t his more efficient and simpler if we take this approach instead.